pandas/core/reshape.py

```diff
@@ -202,10 +202,13 @@ def get_new_columns(self):
                 return self.removed_level
 
             lev = self.removed_level
-            vals = np.insert(lev.astype('object'), 0,
+            vals = np.insert(lev.values.astype('object'), 0,
                              _get_na_value(lev.dtype.type))
 
-            return lev._shallow_copy(vals)
+            try:
+                return lev._shallow_copy(vals.astype(lev.dtype, copy=False))
+            except ValueError:
```

Comment:
I think these are really close in efficiency. But this is obfusciating the code. Either put this try: except: in another function, or just call the index constructor. We don't need another way of doing things.